torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1227,7 +1227,21 @@ std::string ProcessGroupNCCL::getNCCLWatchdogDebugInfo() {
   return retrieveDesyncReport(store_, "NCCL", rank_, size_);
 }
 
-void ProcessGroupNCCL::workCleanupLoop() {
+bool ProcessGroupNCCL::shouldAutoDumpWorkTimeOut(WorkNCCL& work) {
```

Comment:
I'm not following what the intent of this function is.

Reply:
Well, I guess there are two different ways:
1. Each rank keeps polling the TCPStore and check whether the flag has been sent. 
2. Check whether one event is old enough that we need to dump the debug info.

Previously I want to go with #1 but looks like #2 is better so that we don't have a lot of traffic to TCPStore?

Reply:
Basically we want to do (1), but with the optimization that we do not poll the TCPStore when we know it is impossible for another rank to have timed out (`last_time_we_removed_something_from_metaworkList - now < watchdog_timeout`)

Reply:
i guess the only exception here could be if we created a sub-pg on some ranks and no equivalent sub-pgs covering other ranks, then other ranks could observe 'nothing is happening' for a while the trainer really is progressing?

Reply:
they could, but it would have to be a long time of nothing happening (like 20 minutes for the timeout), and all that would happen would be that every 2 seconds they would poll.  Once we are talking about a subset of process groups, who stay inactive for 20 minutes, I am not concerned that their polling would affect overall performance.

Reply:
currently monitor time out is just 2 mins and watchdog timeout is 1 sec. Do you think we might need to make the timeout longer? Also, I updated the code based on my understanding of all comments and feedback. Do you mind taking a look again?

I will do the rebase later (it should be straightforward). And for test, I am still thinking how we can test this case? I mean this changes is just to further ensure we dump the flight recorder and it's kind of hard to mimic the situation when there is no add or erase.. Maybe I should make rank 1 sleep for a while?

Reply:
actually, i think the test i added for dumping on timeout is almost exactly what you need.  You can just customize it slightly so that instead of expecting rank1 to exit cleanly, you expect it also crashes and produces a separate flightrecorder file.  You'll also have to make sure rank1 stays alive long enough to get killed.

https://github.com/pytorch/pytorch/pull/115176/files#diff-9f639571a250cffbe9cded7d2fbb5ad6311e4be9c0c7610e5ba85930806e7f38

Reply:
You could probably make rank1 wait for a signal from the parent which it will never recv.